api/src/pcapi/core/offers/repository.py

```diff
@@ -129,7 +129,12 @@ def get_offers_by_filters(
         search = name_keywords_or_isbn
         if len(name_keywords_or_isbn) > 3:
             search = "%{}%".format(name_keywords_or_isbn)
-        query = query.filter(or_(Offer.name.ilike(search), Offer.extraData["isbn"].astext == name_keywords_or_isbn))
+        # We should really be using `union` instead of `union_all` here since we don't want duplicates but
+        # 1. it's unlikely that a book will contain its ISBN in its name
+        # 2. we need to migrate Offer.extraData to JSONB in order to use `union`
+        query = query.filter(Offer.name.ilike(search)).union_all(
+            query.filter(Offer.extraData["isbn"].astext == name_keywords_or_isbn)
```

Comment:
L'ISBN étant [normalisé](https://www.afnil.org/autre-media/) à une séquence de 13 chiffres commençant par 978 ou 979 pour les livres parus depuis 2007 (et 10 chiffres pour ceux parus avant 2007, mais seuls 20 livres synchronisés ont un isbn de longueur différente de 13¹), on pourrait s'éviter une union en déterminant si l'on recherche sur le nom ou l'ISBN.

Cela demanderait d'implémenter un contrôle identique de l'ISBN dans tous les canaux de création d'offres de livres, possiblement via une contrainte en DB en plus de contrôle de champ sur pro, et dans l'api, mais c'est un vaste chantier en soi !

¹ je remets ici la syntaxe pour aider mon moi futur amnésique:
`SELECT COUNT(*) FROM product WHERE LENGTH( "extraData" ->> 'isbn') != 13 AND "lastProviderId" IS NOT NULL`

Reply:
interessant, très bonne idée